File: results/_analyze.py
import csv, statistics as st
from collections import defaultdict

rows = list(csv.DictReader(open('C:/ConflictualMAS/results/episodes_seed42.csv')))
print(f'Total rows: {len(rows)}')

def num(s):
    try: return float(s)
    except: return None

# Structure: per (mode, city) aggregate over rounds
by_mc = defaultdict(list)
for r in rows:
    by_mc[(r['policy_mode'], r['city'])].append(r)

print('\n=== Mode x City counts ===')
for k in sorted(by_mc): print(f'  {k}: {len(by_mc[k])} eps')

# Headline metrics per mode x city
print('\n=== Headline (mean over all 80 rounds) ===')
cols = ['throughput_rate','accept_rate','completion_per_accepted','latency_mean',
        'agent_completed_gini','agent_completed_std',
        'mean_imp_accepted','mean_imp_refused',
        'accept_rate_high_cong','accept_rate_low_cong',
        'mean_congestion_at_decision','n_ghost_active_mean',
        'mean_extra_steps_per_task','wallclock_ms',
        'actor_loss','entropy']
hdr = f"{'Mode':<16}{'City':<22}" + ''.join(f'{c[:14]:>15}' for c in cols)
print(hdr)
for k in sorted(by_mc):
    line = f"{k[0]:<16}{k[1]:<22}"
    for c in cols:
        vals = [num(r[c]) for r in by_mc[k] if num(r[c]) is not None]
        m = st.mean(vals) if vals else 0
        line += f"{m:>15.3f}"
    print(line)

# Convergence: split into early (rounds 0-19) vs late (rounds 60-79)
print('\n=== Convergence (early rounds 0-19 vs late 60-79) ===')
def slice_rounds(rows_mc, lo, hi):
    # global_episode increases over training but rounds are striped across cities/modes.
    # Use the first half / last quarter by record index instead.
    n = len(rows_mc)
    return rows_mc[int(lo/80*n):int(hi/80*n)]

for k in sorted(by_mc):
    early = slice_rounds(by_mc[k], 0, 20)
    late  = slice_rounds(by_mc[k], 60, 80)
    def m(rows, col):
        vs = [num(r[col]) for r in rows if num(r[col]) is not None]
        return st.mean(vs) if vs else 0
    print(f'  {k[0]:<16}{k[1]:<22}'
          f'  thr early={m(early,"throughput_rate"):.3f} late={m(late,"throughput_rate"):.3f}'
          f'  acc early={m(early,"accept_rate"):.3f} late={m(late,"accept_rate"):.3f}'
          f'  c/a early={m(early,"completion_per_accepted"):.3f} late={m(late,"completion_per_accepted"):.3f}'
          f'  aloss early={m(early,"actor_loss"):.4f} late={m(late,"actor_loss"):.4f}')

# Profile distribution
print('\n=== Congestion profile distribution ===')
prof_counts = defaultdict(int)
for r in rows:
    prof_counts[r['congestion_profile']] += 1
total = sum(prof_counts.values())
for p,c in sorted(prof_counts.items()):
    print(f'  {p or "(empty)":<15} {c:4d}  {c/total*100:.1f}%')

# Selectivity sanity: imp_accepted vs imp_refused
print('\n=== Selectivity (mean imp acc vs ref) per mode ===')
for mode in sorted(set(r['policy_mode'] for r in rows)):
    mode_rows = [r for r in rows if r['policy_mode']==mode]
    accs = [num(r['mean_imp_accepted']) for r in mode_rows if num(r['mean_imp_accepted'])]
    refs = [num(r['mean_imp_refused']) for r in mode_rows if num(r['mean_imp_refused'])]
    high = [num(r['accept_rate_high_cong']) for r in mode_rows if num(r['accept_rate_high_cong'])]
    low  = [num(r['accept_rate_low_cong'])  for r in mode_rows if num(r['accept_rate_low_cong'])]
    print(f'  {mode:<16}'
          f' imp_acc={st.mean(accs):.3f}'
          f' imp_ref={st.mean(refs) if refs else 0:.3f}'
          f'  delta={st.mean(accs)-(st.mean(refs) if refs else 0):+.3f}'
          f'  acc_high_cong={st.mean(high) if high else 0:.3f}'
          f'  acc_low_cong={st.mean(low) if low else 0:.3f}')

File: results/Archives/_analyze.pl
#!/usr/bin/perl
use strict; use warnings;

my @files = ('C:/ConflictualMAS/results/episodes_seed42.csv',
             'C:/ConflictualMAS/results/episodes_seed43.csv');

my @rows;
my %ci;
for my $file (@files) {
    open my $fh, '<', $file or die "$file: $!";
    my $hdr = <$fh>; $hdr =~ s/[\r\n]+$//;
    my @cols = split /,/, $hdr;
    %ci = (); $ci{$cols[$_]} = $_ for 0..$#cols;
    while (my $l = <$fh>) {
        $l =~ s/[\r\n]+$//;
        my @f = split /,/, $l;
        push @rows, \@f;
    }
    close $fh;
}
my $n = scalar @rows;
print "Total rows: $n (across 2 seeds)\n";

sub get { my ($r,$c) = @_; return $r->[$ci{$c}]; }
sub mean { my @v = grep { defined && /\d/ } @_; return 0 unless @v; my $s=0; $s+=$_ for @v; return $s/@v; }
sub stddev { my @v = grep { defined && /\d/ } @_; return 0 if @v < 2;
             my $m = mean(@v); my $s=0; for (@v) { $s += ($_-$m)**2 } return sqrt($s/@v); }

# Group by (seed, mode, city)
my %g;
push @{$g{ get($_,'seed').'|'.get($_,'policy_mode').'|'.get($_,'city') }}, $_ for @rows;

print "\n=== Counts per (seed, mode, city) ===\n";
for my $k (sort keys %g) {
    printf "  %-50s %d eps\n", $k, scalar @{$g{$k}};
}

# Profile distribution per seed
print "\n=== Congestion profile distribution per seed ===\n";
for my $seed (42, 43) {
    my %p;
    for my $r (@rows) { next unless get($r,'seed') eq $seed; $p{get($r,'congestion_profile')||'(empty)'}++; }
    my $tot = 0; $tot += $_ for values %p;
    print "  seed=$seed:";
    printf "  %s=%d (%.0f%%)", $_, $p{$_}, $p{$_}/$tot*100 for sort keys %p;
    print "\n";
}

# Headline metrics: aggregate over seeds + cities per mode
my @cols_show = qw(
  throughput_rate accept_rate completion_per_accepted latency_mean
  agent_completed_gini agent_completed_std
  mean_imp_accepted mean_imp_refused
  accept_rate_high_cong accept_rate_low_cong
  mean_congestion mean_congestion_at_decision n_ghost_active_mean
  mean_extra_steps_per_task
  actor_loss entropy wallclock_ms);

print "\n=== Per-mode headline (mean ± std across 2 seeds × 3 cities × 80 rounds) ===\n";
printf "%-18s", "Metric";
printf " %14s ± %-6s", "MAPPO", "std";
printf " %14s ± %-6s", "FaithfulMAPPER", "std";
print "  ΔMAPPO−FM\n";
for my $col (@cols_show) {
    my @mappo = map { get($_, $col) } grep { get($_,'policy_mode') eq 'MAPPO' } @rows;
    my @fmap  = map { get($_, $col) } grep { get($_,'policy_mode') eq 'FaithfulMAPPER' } @rows;
    my $mm = mean(@mappo); my $sm = stddev(@mappo);
    my $mf = mean(@fmap);  my $sf = stddev(@fmap);
    my $delta = $mm - $mf;
    printf "%-18s %14.4f ± %-6.4f %14.4f ± %-6.4f  %+8.4f\n",
        $col, $mm, $sm, $mf, $sf, $delta;
}

# Convergence per mode per seed (first 25% vs last 25% of records)
print "\n=== Convergence per mode (early Q1 vs late Q4) ===\n";
for my $mode ('MAPPO','FaithfulMAPPER') {
    print "  --- $mode ---\n";
    for my $seed (42, 43) {
        for my $city ('Tokyo_Small','Kyoto_Small','LosAngeles_Small') {
            my @r = @{$g{"$seed|$mode|$city"}};
            my @early = @r[0 .. int(@r/4)-1];
            my @late  = @r[int(@r*3/4) .. $#r];
            printf "    seed=%d %-22s thr: %.3f→%.3f  acc: %.3f→%.3f  c/a: %.3f→%.3f  aloss: %.4f→%.4f  entropy: %.3f→%.3f\n",
                $seed, $city,
                mean(map{get($_,'throughput_rate')}@early), mean(map{get($_,'throughput_rate')}@late),
                mean(map{get($_,'accept_rate')}@early),     mean(map{get($_,'accept_rate')}@late),
                mean(map{get($_,'completion_per_accepted')}@early), mean(map{get($_,'completion_per_accepted')}@late),
                mean(map{get($_,'actor_loss')}@early), mean(map{get($_,'actor_loss')}@late),
                mean(map{get($_,'entropy')}@early), mean(map{get($_,'entropy')}@late);
        }
    }
}

# Selectivity: discrimination signal
print "\n=== Selectivity signal per mode (mean across 2 seeds) ===\n";
for my $mode ('MAPPO','FaithfulMAPPER') {
    my @r = grep { get($_,'policy_mode') eq $mode } @rows;
    my $ia = mean(map { get($_, 'mean_imp_accepted') } @r);
    my $ir = mean(map { get($_, 'mean_imp_refused')  } @r);
    my $ah = mean(map { get($_, 'accept_rate_high_cong') } @r);
    my $al = mean(map { get($_, 'accept_rate_low_cong')  } @r);
    printf "  %-16s imp_acc=%.3f imp_ref=%.3f  Δ_imp=%+.3f  | acc_hi_cong=%.3f acc_lo_cong=%.3f  Δ_cong=%+.3f\n",
        $mode, $ia, $ir, $ia-$ir, $ah, $al, $al-$ah;
}

# Cross-seed variance check: do the two seeds agree?
print "\n=== Cross-seed agreement (per mode, mean per seed) ===\n";
for my $mode ('MAPPO','FaithfulMAPPER') {
    for my $col (qw(throughput_rate accept_rate completion_per_accepted entropy)) {
        my @s42 = map{get($_,$col)} grep { get($_,'policy_mode') eq $mode && get($_,'seed') eq '42' } @rows;
        my @s43 = map{get($_,$col)} grep { get($_,'policy_mode') eq $mode && get($_,'seed') eq '43' } @rows;
        printf "  %-16s %-25s seed42=%.4f  seed43=%.4f  diff=%+.4f\n",
            $mode, $col, mean(@s42), mean(@s43), mean(@s42)-mean(@s43);
    }
}
